File: guardian_sdk/telemetry.py
from __future__ import annotations

import json
import os
import socket
import threading
from dataclasses import asdict, dataclass, field
from datetime import datetime, timezone
from typing import Any, Dict, Optional

import requests

from .config import GuardianConfig


_write_lock = threading.Lock()


def _utc_iso() -> str:
    return datetime.now(timezone.utc).isoformat(timespec="seconds")


@dataclass(frozen=True)
class TelemetryEvent:
    ts: str
    service: str
    env: str
    event_type: str
    message: str
    severity: int = 1
    category: Optional[str] = None
    verdict: Optional[str] = None
    pid: int = field(default_factory=os.getpid)
    host: str = field(default_factory=socket.gethostname)
    extra: Dict[str, Any] = field(default_factory=dict)


class FileTelemetrySink:
    def __init__(self, config: GuardianConfig):
        self._config = config

    def update_config(self, config: GuardianConfig) -> None:
        self._config = config

    def emit(self, ev: TelemetryEvent) -> None:
        line = self.format_line(ev)
        path = self._config.log_path or "security.log"
        try:
            with _write_lock:
                with open(path, "a", encoding="utf-8", errors="replace", newline="\n") as f:
                    f.write(line + "\n")
        except OSError:
            return

    def format_line(self, ev: TelemetryEvent) -> str:
        payload = asdict(ev)
        json_blob = json.dumps(payload, ensure_ascii=False, separators=(",", ":"))
        date_prefix = ev.ts[:10] if ev.ts else _utc_iso()[:10]

        core = (
            f"{date_prefix} guardian event_type={ev.event_type} "
            f"service={ev.service} env={ev.env} "
            f"severity={ev.severity} "
            f"message={ev.message}"
        )
        if ev.category:
            core += f" category={ev.category}"
        if ev.verdict:
            core += f" verdict={ev.verdict}"
        return core + " | " + json_blob


class RemoteTelemetrySink:
    def __init__(self, config: GuardianConfig):
        self._config = config

    def update_config(self, config: GuardianConfig) -> None:
        self._config = config

    def emit(self, ev: TelemetryEvent) -> None:
        if not self._config.ingest_url:
            return

        headers = {
            "Content-Type": "application/json",
            "X-API-KEY": os.getenv("GUARDIAN_API_KEY", ""),
        }
        payload = asdict(ev)

        try:
            requests.post(
                self._config.ingest_url,
                json=payload,
                headers=headers,
                timeout=1,
            )
        except Exception:
            return


class Telemetry:
    def __init__(self, config: GuardianConfig):
        self._config = config
        self._file_sink = FileTelemetrySink(config)
        self._remote_sink = RemoteTelemetrySink(config)

    def update_config(self, config: GuardianConfig) -> None:
        self._config = config
        self._file_sink.update_config(config)
        self._remote_sink.update_config(config)

    def emit(
        self,
        *,
        event_type: str,
        message: str,
        severity: int = 1,
        category: Optional[str] = None,
        verdict: Optional[str] = None,
        extra: Optional[Dict[str, Any]] = None,
    ) -> None:
        if not self._config.enabled:
            return

        ev = TelemetryEvent(
            ts=_utc_iso(),
            service=self._config.service_name,
            env=self._config.environment,
            event_type=event_type,
            message=message,
            severity=int(severity),
            category=category,
            verdict=verdict,
            extra=extra or {},
        )
        self._file_sink.emit(ev)
        self._remote_sink.emit(ev)

File: guardian-sdk-node/src/telemetry.js
const fs = require('fs');
const http = require('http');
const https = require('https');
const os = require('os');
const { URL } = require('url');

function utcIso() {
    return new Date().toISOString().replace(/\.[0-9]{3}Z$/, '+00:00'); // Close enough to python's isoformat with seconds
}

class TelemetryEvent {
    constructor({
        ts, service, env, event_type, message, severity = 1,
        category = null, verdict = null, pid = process.pid,
        host = os.hostname(), extra = {}
    }) {
        this.ts = ts;
        this.service = service;
        this.env = env;
        this.event_type = event_type;
        this.message = message;
        this.severity = severity;
        this.category = category;
        this.verdict = verdict;
        this.pid = pid;
        this.host = host;
        this.extra = extra;
    }
}

class FileTelemetrySink {
    constructor(config) {
        this.config = config;
    }

    emit(ev) {
        const line = this.formatLine(ev);
        const path = this.config.logPath || "security.log";
        try {
            fs.appendFileSync(path, line + '\n', { encoding: 'utf-8' });
        } catch (err) {
            // Never break app because logging failed
        }
    }

    formatLine(ev) {
        const payload = { ...ev };
        const jsonBlob = JSON.stringify(payload);
        
        const datePrefix = ev.ts ? ev.ts.substring(0, 10) : new Date().toISOString().substring(0, 10);
        
        let core = `${datePrefix} guardian event_type=${ev.event_type} service=${ev.service} env=${ev.env} severity=${ev.severity} message=${ev.message}`;
        if (ev.category) core += ` category=${ev.category}`;
        if (ev.verdict) core += ` verdict=${ev.verdict}`;
        
        return core + ' | ' + jsonBlob;
    }
}

class RemoteTelemetrySink {
    constructor(config) {
        this.config = config;
    }

    emit(ev) {
        const ingestUrl = this.config.ingestUrl;
        if (!ingestUrl) {
            return;
        }

        const apiKey = process.env.GUARDIAN_API_KEY || '';
        const payload = JSON.stringify({ ...ev });

        setImmediate(() => {
            try {
                const url = new URL(ingestUrl);
                const client = url.protocol === 'https:' ? https : http;
                const req = client.request(
                    {
                        method: 'POST',
                        hostname: url.hostname,
                        port: url.port || (url.protocol === 'https:' ? 443 : 80),
                        path: `${url.pathname}${url.search}`,
                        headers: {
                            'Content-Type': 'application/json',
                            'Content-Length': Buffer.byteLength(payload),
                            'X-API-KEY': apiKey
                        },
                        timeout: 1000
                    },
                    (res) => {
                        res.resume();
                    }
                );

                req.on('error', () => {});
                req.on('timeout', () => {
                    req.destroy();
                });
                req.write(payload);
                req.end();
            } catch (err) {
                // Never break app because remote telemetry failed.
            }
        });
    }
}

class Telemetry {
    constructor(config) {
        this.config = config;
        this.fileSink = new FileTelemetrySink(config);
        this.remoteSink = new RemoteTelemetrySink(config);
    }

    emit({ event_type, message, severity = 1, category = null, verdict = null, extra = {} }) {
        if (!this.config.enabled) return;

        const ev = new TelemetryEvent({
            ts: new Date().toISOString(),
            service: this.config.serviceName,
            env: this.config.environment,
            event_type,
            message,
            severity: parseInt(severity, 10),
            category,
            verdict,
            extra
        });
        this.fileSink.emit(ev);
        this.remoteSink.emit(ev);
    }
}

module.exports = { Telemetry, TelemetryEvent, FileTelemetrySink, RemoteTelemetrySink };
